dialogue-core/src/main/java/com/palantir/dialogue/core/UserAgentChannel.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.dialogue.core;
+
+import com.google.common.util.concurrent.ListenableFuture;
+import com.palantir.conjure.java.api.config.service.UserAgent;
+import com.palantir.conjure.java.api.config.service.UserAgents;
+import com.palantir.dialogue.Channel;
+import com.palantir.dialogue.Endpoint;
+import com.palantir.dialogue.Request;
+import com.palantir.dialogue.Response;
+
+/**
+ * Adds a {@code user-agent} header that is the combination of the given base user agent, the version of the
+ * dialogue library (extracted from this package's implementation version), and the name and version of the
+ * {@link Endpoint}'s target service and endpoint.
+ */
+final class UserAgentChannel implements Channel {
+
+    private final Channel delegate;
+    private final UserAgent baseAgent;
+
+    UserAgentChannel(Channel delegate, UserAgent baseAgent) {
+        this.delegate = delegate;
+        this.baseAgent = baseAgent;
+    }
+
+    @Override
+    public ListenableFuture<Response> execute(Endpoint endpoint, Request request) {
+        Request newRequest = Request.builder()
+                .from(request)
+                .putHeaderParams("user-agent", UserAgents.format(augmentUserAgent(baseAgent, endpoint)))
+                .build();
+        return delegate.execute(endpoint, newRequest);
+    }
+
+    private static UserAgent augmentUserAgent(UserAgent baseAgent, Endpoint endpoint) {
+        String maybeDialogueVersion = Channel.class.getPackage().getImplementationVersion();
+        UserAgent.Agent dialogueAgent = UserAgent.Agent.of(
+                "dialogue",
+                maybeDialogueVersion != null ? maybeDialogueVersion : "0.0.0");
+
+        return baseAgent
+                .addAgent(UserAgent.Agent.of(endpoint.serviceName(), endpoint.version()))
+                .addAgent(UserAgent.Agent.of(endpoint.endpointName(), endpoint.version()))
```

Comment:
let's take this out for now. in particular because it's nice to aggregate by user agent and see who are the biggest callers of a particular service. you're also able to sub-aggregate by path (eg: endpoint) at that point anyway.

Reply:
done